**Q1. What is an LLM, and what implications does it have for economics or finance?**

LLM은 트랜스포머 아키텍처를 기반으로 방대한 데이터를 학습하여 인간의 언어를 이해하고 추론하는 딥러닝 모델로, 금융 및 경제 분야에서 비정형 데이터의 활용 방식을 근본적으로 혁신하고 있습니다. 과거 전통적인 계량 경제 모형(OLS 회귀, ARIMA 시계열 분석 등)이 주로 과거의 주가, 거래량, 금리와 같이 이미 수치화된 정형 데이터 간의 선형적 관계를 추정하는 데 국한되었던 반면, LLM은 Self-Attention 메커니즘을 통해 단어 간의 문맥과 미묘한 뉘앙스까지 파악하여 자연어를 고차원 벡터 공간에 임베딩합니다. 이러한 인지 및 추론 로직을 통해 LLM은 연준 의사록의 매파적/비둘기파적 어조 변화, 기업 실적 발표에서 경영진이 사용하는 단어의 긍정/부정 비율, 실시간 글로벌 금융 뉴스 등 방대한 텍스트 데이터 속에서 시장의 감성과 숨겨진 알파 시그널을 즉각적인 수치로 변환하고 계량화할 수 있습니다. 이는 과거 거대 자본을 가진 기관 투자자들만이 전문 인력을 동원해 수행할 수 있었던 방대한 정성적 정보의 분석 비용과 시간을 획기적으로 단축시켜 새로운 텍스트 정보가 자산 가격에 즉시 반영되도록 이끌며, 결과적으로 개인과 기관 간의 정보 비대칭성을 크게 해소하고 시장의 준강형 효율성을 제고합니다. 나아가, 이렇게 정량화된 텍스트 데이터는 자산 가격 책정 모형의 독립적인 새로운 변수로 추가되어 전통적 재무 지표가 설명하지 못하는 시장의 심리적 움직임을 포착하게 해 주므로, 자연어 처리(NLP) 기술을 알고리즘 트레이딩에 직접 융합하여 기존의 수치 기반 모델을 뛰어넘는 차세대 퀀트 전략의 고도화를 촉진하는 중대한 파급 효과를 지닙니다.

**Q2. What are vibe coding and harness engineering, and what implications do they have for economics or finance?**

Vibe Coding은 엄격한 프로그래밍 문법 대신 자연어 프롬프트를 통해 개발자의 직관적인 의도만으로 AI가 코드를 생성하게 하는 패러다임이며, Harness Engineering은 이렇게 생성된 알고리즘의 성능, 안전성, 편향성을 체계적으로 검증하고 통제하는 평가 프레임워크를 구축하는 작업입니다. 경제 및 금융 산업에서 바이브 코딩은 프로그래밍 진입 장벽을 완전히 허물어, 단순한 코딩 구현 능력보다 투자자의 경제학적 직관과 도메인 지식 자체를 시장의 핵심 경쟁 우위로 격상시키는 중대한 영향을 미칩니다. 이를 통해 금융 분석가들은 막대한 IT 리서치 비용을 절감하면서도 딥러닝 기반의 퀀트 모델이나 알고리즘 트레이딩 전략을 신속하게 프로토타이핑하여 새로운 알파 창출 아이디어의 시장 적용 속도를 극대화할 수 있습니다. 그러나 금융 시장은 미세한 알고리즘 오류나 AI의 환각(Hallucination) 현상이 대규모 연쇄 청산이나 막대한 자본 손실과 같은 시스템적 리스크로 직결되므로 하니스 엔지니어링이 반드시 동반되어야 합니다. 하니스 엔지니어링은 엄격한 스트레스 테스트와 백테스팅을 통해 모델의 과적합을 방지하고, 금융 규제 당국이 요구하는 설명 가능한 AI(Explainable AI) 기준을 충족시키는 강력한 통제 수단으로 작용합니다. 결론적으로 바이브 코딩이 금융 혁신의 비용을 낮추고 민첩성을 높이는 '가속 페달'이라면, 하니스 엔지니어링은 리스크를 관리하고 시장의 신뢰를 방어하는 안전 장치로 기능하며, 이 둘의 상호보완적 결합이야말로 차세대 금융 시스템이 극한의 변동성 속에서도 안정적으로 작동하기 위한 필수 불가결한 인프라가 됩니다.

**Q3. In what ways does wiki-LLM improve on traditional RAG systems?**

wiki-LLM은 문서를 기계적으로 분할하여 단순한 텍스트 임베딩 간의 벡터 유사도에만 의존하는 전통적인 RAG 시스템의 정보 파편화 문제를 해결함으로써 문서 탐색 및 추론의 품질을 혁신적으로 향상시킵니다. 기존의 전통적 RAG는 연준 의사록이나 방대한 산업 리포트에서 단순히 키워드와 유사한 문장 조각들을 문맥 없이 추출하여 LLM에 주입하기 때문에, 문서 전체를 아우르는 포괄적 문맥이 소실되는 치명적인 한계가 있었습니다. 반면 wiki-LLM은 위키피디아의 하이퍼링크 구조처럼 개별 개념들 간의 계층적 관계와 의미론적 연결망을 지식 그래프 형태로 사전 구축하는 방식을 취합니다. 이러한 구조적 향상은 경제 및 금융 분야에서 필수적인 다중 홉 추론(Multi-hop reasoning), 즉 '연준의 금리 인하(A)'가 '특정 원자재 밸류체인(B)'을 거쳐 최종적으로 '개별 기업의 밸류에이션(C)'에 미치는 복잡한 연쇄적 파급 효과를 맥락의 끊김 없이 유기적으로 추적하는 것을 가능하게 합니다. 결과적으로 wiki-LLM은 단순 텍스트 짜깁기로 인해 발생하는 AI의 환각 현상을 획기적으로 억제할 뿐만 아니라, 금융 규제 및 실무에서 가장 중요한 '정보 출처의 명확한 추적성'을 보장하여 보다 신뢰도 높고 입체적인 퀀트 리서치 시스템을 구축하는 핵심 기반으로 작용합니다.

# Q4. RAG Lab 재현 과제

**학번:** 20220727 | **이름:** 조현모 | **과목:** AI Finance (2026 Spring)

---

## 과제 목표

> **Q4. Replication task:** Recreate the class GitHub directory below using your own documents or economics/finance papers:  
> https://github.com/cschung7/2026-spring-gse-AIFinance/tree/main/rag-lab/rag_lab

수업 GitHub의 `rag-lab/rag_lab` 구조를 직접 선택한 경제·금융 논문 3편으로 재현한다.  
**Standard RAG**와 **Karpathy LLM-Wiki** 두 방식을 구현하고, 4가지 질의로 작동을 검증한다.

---

## 사용 문서 (data/ 폴더)

| 파일 | 논문/문서 | 선택 이유 |
|------|----------|-----------|
| `empircial asset pricing via macine learning.pdf` | Gu, Kelly & Xiu (2020) | ML이 자산가격 예측에 미치는 핵심 연구 |
| `Temporal Fusion Transformers...pdf` | Lim et al. (2021) | 트랜스포머 기반 금융 시계열 예측 대표 논문 |
| `fomcminutes20260318.pdf` | FOMC 의사록 (2026.03.18) | 실제 중앙은행 문서로 RAG 실용성 검증 |

In [1]:
import os
import sys
import textwrap
from pypdf import PdfReader

sys.path.append(os.path.abspath('..'))
from rag_lab.embeddings import EmbeddingEngine
from rag_lab.standard_rag import StandardRAG

# --- Groq LLM으로 교체 (원본 llm.py 보존, llm_groq.py 사용) ---
import rag_lab.standard_rag as _sr
import rag_lab.karpathy_wiki as _kw
from rag_lab.llm_groq import llm_call, llm_call_json
_sr.llm_call = llm_call
_kw.llm_call = llm_call
_kw.llm_call_json = llm_call_json
print("✅ LLM 백엔드: Groq (llama-3.3-70b-versatile)")

# --- PDF 로드 ---
data_dir = "./data"
papers_dict = {}

print("\n[Step 1] PDF 문서 텍스트 추출 시작...")
for filename in sorted(os.listdir(data_dir)):
    if filename.endswith(".pdf"):
        reader = PdfReader(os.path.join(data_dir, filename))
        text = "".join(p.extract_text() or "" for p in reader.pages)
        papers_dict[filename] = text
        print(f"  ✔ '{filename}'  ({len(text):,}자)")

# --- StandardRAG 구축 ---
print("\n[Step 2] 임베딩 엔진 초기화 및 FAISS 인덱스 구축 중...")
engine = EmbeddingEngine()
my_rag = StandardRAG(papers=papers_dict, engine=engine)
print(f"\n✅ 준비 완료 | 문서 {len(papers_dict)}편 | 청크 {len(my_rag.chunks)}개 | FAISS {my_rag.index.ntotal}벡터")

/mnt/c/chm26y/학교/AI금융/중간고사/2026-spring-gse-AIFinance/rag-lab/rag_lab/venv/lib/python3.11/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


✅ LLM 백엔드: Groq (llama-3.3-70b-versatile)

[Step 1] PDF 문서 텍스트 추출 시작...
  ✔ 'Temporal Fusion Transformers for Interpretable Multi-horizon Time Series Forecasting.pdf'  (63,278자)
  ✔ 'empircial asset pricing via macine learning.pdf'  (211,651자)
  ✔ 'fomcminutes20260318.pdf'  (41,747자)

[Step 2] 임베딩 엔진 초기화 및 FAISS 인덱스 구축 중...
  임베딩 모델 로딩: all-MiniLM-L6-v2


Loading weights: 100%|██████████| 103/103 [00:00<00:00, 2291.96it/s]
/mnt/c/chm26y/학교/AI금융/중간고사/2026-spring-gse-AIFinance/rag-lab/rag_lab/embeddings.py:27: FutureWarning: The `get_sentence_embedding_dimension` method has been renamed to `get_embedding_dimension`.
  self.dim = self.model.get_sentence_embedding_dimension()


  → 차원 수: 384
[Standard RAG] 3편 논문 청킹 중...
  → 178개 청크 생성
[Standard RAG] 임베딩 생성 중...


Batches: 100%|██████████| 6/6 [00:05<00:00,  1.07it/s]

  → FAISS 인덱스 구축 완료 (178 벡터)

✅ 준비 완료 | 문서 3편 | 청크 178개 | FAISS 178벡터


---

> **[LLM 백엔드 변경 안내]**
>
> 교수님이 제공하신 `llm.py`는 OpenAI API(`gpt-4o-mini`)를 기반으로 작성되어 있습니다.  
> 그러나 본 과제 수행 시 해당 OpenAI 계정의 **크레딧 소진(429 quota exceeded)** 으로 인해 LLM 호출이 불가능했습니다.
>
> 이에 원본 `llm.py`는 **수정 없이 그대로 보존**하고,  
> 동일한 인터페이스(`llm_call`, `llm_call_json`)를 유지하는 **`llm_groq.py`를 별도로 작성**하여  
> [Groq](https://groq.com)의 무료 API(`llama-3.3-70b-versatile` 모델)로 교체하였습니다.
>
> | 항목 | 원본 (`llm.py`) | 대체 (`llm_groq.py`) |
> |------|----------------|---------------------|
> | 백엔드 | OpenAI API | Groq API |
> | 모델 | gpt-4o-mini | llama-3.3-70b-versatile |
> | 인터페이스 | `llm_call`, `llm_call_json` | 동일 |
> | 원본 보존 | — | ✅ llm.py 수정 없음 |
>
> 교체 방식: Setup 셀에서 `rag_lab.standard_rag`와 `rag_lab.karpathy_wiki` 모듈의  
> `llm_call` 참조를 런타임에 Groq 버전으로 교체(monkey-patch)합니다.

---

## Method 1 — Standard RAG: 4가지 질의로 검색 검증

각 논문을 타깃으로 한 영어 질문 3개 + 논문 간 교차 질문 1개를 사용한다.  
영어로 질문하는 이유: 임베딩 모델(`all-MiniLM-L6-v2`)이 영어 최적화 모델이므로  
영어 질문 시 유사도 점수가 유의미하게 높아진다.

| 질문 | 대상 문서 |
|------|-----------|
| Q1: ML models for stock return prediction | Gu et al. (2020) |
| Q2: TFT architecture and variable selection | Lim et al. (2021) |
| Q3: Fed's rate decision and economic outlook (March 2026) | FOMC 의사록 |
| Q4: Cross-paper — neural networks vs traditional methods in finance | 교차 질문 |

In [2]:
# 4가지 질의 정의
queries = [
    {
        "id": "Q1",
        "target": "Gu et al. (2020)",
        "text": "What machine learning models are used to predict stock returns, "
                "and how do they compare to OLS regression in terms of performance?"
    },
    {
        "id": "Q2",
        "target": "Lim et al. (2021) — TFT",
        "text": "What is the architecture of the Temporal Fusion Transformer "
                "and how does it perform variable selection for time series forecasting?"
    },
    {
        "id": "Q3",
        "target": "FOMC Minutes (2026-03)",
        "text": "What was the Federal Reserve's decision on interest rates and "
                "what were the main economic risks discussed in the March 2026 FOMC meeting?"
    },
    {
        "id": "Q4",
        "target": "Cross-paper",
        "text": "How do neural networks and attention-based models improve "
                "financial forecasting compared to traditional linear methods?"
    },
]

# retrieve_only 실행 (LLM 불필요)
all_results = {}
for q in queries:
    all_results[q["id"]] = my_rag.retrieve_only(q["text"], top_k=3)

# 결과 출력 (청크 텍스트 500자까지 표시)
for q in queries:
    chunks = all_results[q["id"]]
    print("=" * 80)
    print(f"[{q['id']}] Target: {q['target']}")
    print(f"Query : {q['text']}")
    print("-" * 80)
    for i, item in enumerate(chunks):
        preview = item['chunk'].replace('\n', ' ').strip()[:500]
        print(f"  Rank {i+1} | score={item['score']:.4f} | {item['source'][:55]}")
        print(f"  └ {textwrap.fill(preview, width=76, subsequent_indent='    ')}")
    print()

[Q1] Target: Gu et al. (2020)
Query : What machine learning models are used to predict stock returns, and how do they compare to OLS regression in terms of performance?
--------------------------------------------------------------------------------
  Rank 1 | score=0.6343 | empircial asset pricing via macine learning.pdf
  └ features. It is parsimonious and simple, and comparing against this
    benchmark is conservative because it is highly selected (the
    characteristics it includes are routinely demonstrated to be among the
    most robust return predictors). Lewellen (2015) demonstrates that this
    model performs about as well as larger and more complex stock prediction
    models studied in the literature. In our sample, which is longer and
    wider (more observations in terms of both dates and stocks) than that
    studied in Lewell
  Rank 2 | score=0.5812 | empircial asset pricing via macine learning.pdf
  └ resentative of predictive analytics tools from various branches o

In [3]:
# Q1~Q4 전체 Rank·Source 종합 요약표
src_short = {
    "empircial asset pricing via macine learning.pdf": "Gu et al. (2020)",
    "Temporal Fusion Transformers for Interpretable Multi-horizon Time Series Forecasting.pdf": "Lim et al. (2021)",
    "fomcminutes20260318.pdf": "FOMC Minutes",
}

print("=" * 72)
print("[종합] Q1–Q4 retrieve_only 검색 결과 요약")
print("=" * 72)

for q in queries:
    print(f"\n  [{q['id']}] {q['target']}")
    q_text = q['text']
    print(f"  Query: {q_text[:72]}..." if len(q_text) > 72 else f"  Query: {q_text}")
    print(f"  {'Rank':>6}  {'Score':>7}  출처")
    print("  " + "-" * 50)
    for i, item in enumerate(all_results[q["id"]]):
        src = src_short.get(item['source'], item['source'][:30])
        print(f"  {'Rank '+str(i+1):>6}  {item['score']:>7.4f}  {src}")

print()
print("=" * 72)
print("[패턴 분석] 질문별 검색 출처 분포")
print("=" * 72)
print(f"  {'질문':<6}  {'Top Score':>9}  검색 출처 분포")
print("  " + "-" * 55)
for q in queries:
    chunks = all_results[q["id"]]
    sources = [src_short.get(c['source'], c['source'][:20]) for c in chunks]
    unique = list(dict.fromkeys(sources))
    top_score = chunks[0]['score']
    pattern = "단일 논문" if len(unique) == 1 else f"교차 ({', '.join(unique)})"
    print(f"  {q['id']:<6}  {top_score:>9.4f}  {pattern}")

[종합] Q1–Q4 retrieve_only 검색 결과 요약

  [Q1] Gu et al. (2020)
  Query: What machine learning models are used to predict stock returns, and how ...
    Rank    Score  출처
  --------------------------------------------------
  Rank 1   0.6343  Gu et al. (2020)
  Rank 2   0.5812  Gu et al. (2020)
  Rank 3   0.5529  Gu et al. (2020)

  [Q2] Lim et al. (2021) — TFT
  Query: What is the architecture of the Temporal Fusion Transformer and how does...
    Rank    Score  출처
  --------------------------------------------------
  Rank 1   0.5669  Lim et al. (2021)
  Rank 2   0.5232  Lim et al. (2021)
  Rank 3   0.5203  Lim et al. (2021)

  [Q3] FOMC Minutes (2026-03)
  Query: What was the Federal Reserve's decision on interest rates and what were ...
    Rank    Score  출처
  --------------------------------------------------
  Rank 1   0.6582  FOMC Minutes
  Rank 2   0.6346  FOMC Minutes
  Rank 3   0.6282  FOMC Minutes

  [Q4] Cross-paper
  Query: How do neural networks and attention-based models impr

### retrieve_only 결과 해석 (실제 실행 결과)

**[Q1] Gu et al. (2020) — ML models vs OLS**
- Rank 1~3 모두 `empircial asset pricing via macine learning.pdf` (score: 0.6343 / 0.5812 / 0.5529)
- **Rank 1**: OLS 벤치마크 비교 구간. "월 단위 패널 R²=0.16%"라는 수치가 포함된 절로,
  질문의 핵심(ML vs OLS 성능 비교)과 직접 일치하는 청크가 Rank 1에 올라옴.
- **Rank 2**: 연구 범위 설명 구간. 30,000개 주식·60년(1957~2016)·94개 특성·900개 이상 시그널이 언급되어
  질문의 "어떤 모델을 사용했나"에 관련된 맥락 제공.
- **Rank 3**: 저자 소속 및 Introduction 도입부. 직접적 답변 내용은 적으나,
  논문 전체의 핵심 주장("머신러닝 방법론 비교 분석")이 담긴 구간.

**[Q2] Lim et al. (2021) — TFT 아키텍처**
- Rank 1~3 모두 `Temporal Fusion Transformers...pdf` (score: 0.5669 / 0.5232 / 0.5203)
- **Rank 1**: 기존 Transformer를 금융 시계열에 바로 적용할 수 없는 이유 설명 구간.
  "다양한 유형의 입력(정적·동적·외생 변수)을 구분하지 못한다"는 TFT의 핵심 동기가 담겨 있음.
- **Rank 2**: 논문 Abstract. static covariates·known future inputs·exogenous time series의
  복잡한 혼합이 과제임을 명시 — 질문의 "variable selection" 키워드와 잘 매칭됨.
- **Rank 3**: Transformer 블록 내부 구조 설명 (Gated Residual Connection, Quantile Output).
  아키텍처 세부 설계가 담긴 구간으로, 질문의 "architecture" 키워드에 직접 대응.

**[Q3] FOMC 의사록 — 금리 결정**
- Rank 1~3 모두 `fomcminutes20260318.pdf` (score: 0.6582 / 0.6346 / 0.6282)
- **가장 높은 유사도(0.65~0.66)**. 정책 문서는 자연어 서술 비중이 높아 임베딩 매칭이 잘 됨.
- **Rank 1**: 회의 개요 (2026년 3월 17~18일 공동 회의 헤더).
- **Rank 2**: 금리 경로 변화. 선물 시장이 금년 내 금리 인하를 더 이상 완전히 반영하지 않으며,
  옵션 기반 modal path도 "올해 금리 동결"로 이동했음을 기술.
- **Rank 3**: 표결 결과. Powell 등 10인 찬성, **Miran 1인 반대**(인하 선호).
  질문의 "decision on interest rates"에 가장 직접적으로 답하는 청크.

**[Q4] 교차 질문 — 신경망 vs 전통 방법론**
- Rank 1~3 모두 `empircial asset pricing via macine learning.pdf` (score: 0.6093 / 0.5675 / 0.5622)
- **교차 논문 검색 실패**: Lim et al.(TFT)나 FOMC 문서는 단 하나도 검색되지 않음.
  이는 Standard RAG의 핵심 한계를 보여준다 —
  질문과 코사인 유사도가 가장 높은 단일 논문(Gu et al.)만 반복 반환.
- **Rank 1**: 참고문헌 목록 구간(Friedman, Glorot 등) — 유용한 답변 내용이 아님.
  청크 단위 검색이 참고문헌 섹션을 본문과 구별하지 못하는 한계.
- **Rank 2**: "신경망(NN3) Sharpe ratio 0.77, buy-and-hold 대비 26%p 우위"라는
  구체적 성과 수치 포함 — 이 질문에 실제로 유용한 내용.
- **Rank 3**: "트리 기반·신경망이 비선형 설정에서 유의하게 우월" + "ML이 할 수 없는 것" 절.

> **Q4 교훈**: 여러 논문을 종합해야 하는 질문은 Standard RAG로 충분한 답변이 어렵다.
> KarpathyWiki는 논문 간 `[[wiki links]]`로 연결된 위키를 탐색하므로 이런 질문에 더 적합하다.

---

## Method 1 (cont.) — Standard RAG: LLM 생성 포함 전체 파이프라인

`retrieve_only`는 검색만 한다. 아래는 검색 + LLM 생성까지 포함하는 `query()`를  
4가지 질문 각각에 실행한 결과다.

In [4]:
# Standard RAG 전체 파이프라인: Retrieve + Generate (LLM 필요)
print("[Method 1] Standard RAG — Retrieve + Generate")
print("=" * 80)

rag_answers = {}
for q in queries:
    print(f"\n[{q['id']}] {q['target']}")
    print(f"Query: {q['text']}")
    print("-" * 80)
    try:
        result = my_rag.query(q["text"], top_k=3)
        rag_answers[q["id"]] = result.answer
        print(f"Sources    : {result.sources}")
        print(f"Time       : retrieve {result.retrieval_time:.2f}s + generate {result.generation_time:.2f}s")
        print(f"Answer     :\n{result.answer}")
    except Exception as e:
        rag_answers[q["id"]] = ""
        print(f"[오류] {type(e).__name__}: {e}")
    print()

[Method 1] Standard RAG — Retrieve + Generate

[Q1] Gu et al. (2020)
Query: What machine learning models are used to predict stock returns, and how do they compare to OLS regression in terms of performance?
--------------------------------------------------------------------------------
Sources    : ['empircial asset pricing via macine learning.pdf']
Time       : retrieve 0.01s + generate 0.00s
Answer     :
machine learning 모델은 주식 수익률을 예측하는 데 사용되며, OLS 회귀와 비교하여 성능이 더 좋습니다. 구체적으로, elastic net, principal components regression (PCR), partial least squares (PLS)와 같은 모델이 OLS 회귀보다 더 좋은 성능을 보입니다. 예를 들어, elastic net은 900개 이상의 예측 변수를 사용하여 OLS 회귀의 효율성을 개선하고, PCR과 PLS는 예측 변수의 차원을 줄여서 더 좋은 성능을 보입니다. 이러한 모델은 OLS 회귀보다 더 높은 out-of-sample R2 값을 보입니다. (출처: empircial asset pricing via macine learning.pdf, p. 1-2)

또한, 뉴럴 네트워크와 같은 machine learning 모델은 주식 수익률을 예측하는 데 사용되어 경제적으로 의미 있는 결과를 얻을 수 있습니다. 예를 들어, 뉴럴 네트워크를 사용하여 S&P 500을 타이밍하는 포트폴리오 전략은 연간화 된 샤피 비율이 0.77로,买하고 홀드하는 투자자의 0.51보다 더 높은 성과를 보입니다. (출처: em

---

## Method 2 — Karpathy LLM-Wiki

### Standard RAG와의 핵심 차이

```
Standard RAG  : 질문 → 원본 청크 검색 → LLM 답변  (지식을 매번 재발견)
Karpathy Wiki : 문서 → LLM이 위키 작성 → 위키 탐색 → LLM 답변  (지식을 한 번 컴파일)
```

### Compile Phase (최초 1회)
논문 3편 → LLM이 각각 요약 `.md` 생성 → 공통 개념 페이지 생성 → `index.md` 작성

### Query Phase
`index.md` → LLM이 관련 페이지 선택 → 위키 페이지 읽기 → 종합 답변 생성

In [5]:
# Method 2: Karpathy LLM-Wiki
from rag_lab.karpathy_wiki import KarpathyWiki
from pathlib import Path

print("[Method 2] Karpathy LLM-Wiki")
print("=" * 80)

try:
    wiki = KarpathyWiki(papers=papers_dict, wiki_dir=Path("./wiki"), force_rebuild=False)

    # 생성된 위키 페이지 목록
    pages = wiki.list_pages()
    print(f"\n위키 페이지 ({len(pages)}개): {pages}")

    # 4가지 질문으로 위키 쿼리
    wiki_answers = {}
    for q in queries:
        print(f"\n[{q['id']}] {q['target']}")
        print(f"Query: {q['text']}")
        print("-" * 80)
        result = wiki.query(q["text"])
        wiki_answers[q["id"]] = result.answer
        print(f"탐색 페이지 수 : {result.pages_navigated}")
        print(f"Time          : retrieve {result.retrieval_time:.2f}s + generate {result.generation_time:.2f}s")
        print(f"Answer        :\n{result.answer}")

    # 위키 건강 검사
    print("\n" + "=" * 80)
    lint = wiki.lint()
    print(f"[Wiki Lint] 총 페이지: {lint['total_pages']} | 고아 페이지: {lint['orphan_pages']} | 깨진 링크: {len(lint['broken_links'])}")

except Exception as e:
    wiki_answers = {}
    print(f"[오류] {type(e).__name__}: {e}")

[Method 2] Karpathy LLM-Wiki
[Karpathy LLM-Wiki] 3편 논문을 위키로 컴파일 중...
  📄 Temporal Fusion Transformers for Interpretable Multi-horizon Time Series Forecasting.pdf 수집 중...
  📄 empircial asset pricing via macine learning.pdf 수집 중...
  📄 fomcminutes20260318.pdf 수집 중...
  🔗 크로스 커팅 개념 페이지 생성 중...
  📋 index.md 생성 중...
  → 위키 컴파일 완료: 13개 페이지

위키 페이지 (12개): ['Temporal Fusion Transformers for Interpretable Multi-horizon Time Series Forecasting.pdf_summary', 'concept_attention_mechanisms', 'concept_deep_learning', 'concept_economic_indicators', 'concept_explainable_ai', 'concept_financial_markets', 'concept_machine_learning', 'concept_monetary_policy', 'concept_time_series_analysis', 'empircial asset pricing via macine learning.pdf_summary', 'fomcminutes20260318.pdf_summary', 'index']

[Q1] Gu et al. (2020)
Query: What machine learning models are used to predict stock returns, and how do they compare to OLS regression in terms of performance?
------------------------------------------------------

---

## 두 방식 비교

In [6]:
# 아키텍처 비교표
print("=" * 80)
print("[비교] Standard RAG  vs.  Karpathy LLM-Wiki")
print("=" * 80)

rows = [
    ("지식 표현",       "원본 청크 (raw text)",       "LLM 정리 위키 (.md)"),
    ("지식 컴파일",     "없음 (매 쿼리 재발견)",      "1회 후 축적·재사용"),
    ("크로스 레퍼런스", "없음",                       "[[wiki links]] 자동 연결"),
    ("구축 비용",       "낮음 (임베딩만)",             "높음 (논문당 LLM 호출)"),
    ("쿼리 속도",       "빠름 (벡터 검색)",           "중간 (index 탐색)"),
    ("확장성",          "매우 높음 (FAISS)",           "중소 규모 적합"),
    ("투명성",          "낮음 (청크 블랙박스)",        "높음 (.md 직접 열람)"),
    ("교차 질문 품질",  "낮음 (단편 청크)",            "높음 (위키 종합)"),
]

w = [18, 26, 26]
print(f"{'항목':<{w[0]}} | {'Standard RAG':<{w[1]}} | {'LLM-Wiki':<{w[2]}}")
print("─" * (sum(w) + 6))
for r in rows:
    print(f"{r[0]:<{w[0]}} | {r[1]:<{w[1]}} | {r[2]:<{w[2]}}")

# 검색 통계 요약
print(f"\n[Standard RAG 검색 통계]")
print(f"  문서 {len(papers_dict)}편 | 청크 {len(my_rag.chunks)}개 | FAISS {my_rag.index.ntotal}벡터")
print()
print(f"  {'질문':<6} {'대상':<24} {'Rank1 Score':>12} {'Rank1 출처':<40}")
print("  " + "-" * 85)
for q in queries:
    top = all_results[q["id"]][0]
    print(f"  {q['id']:<6} {q['target']:<24} {top['score']:>12.4f} {top['source'][:40]:<40}")

[비교] Standard RAG  vs.  Karpathy LLM-Wiki
항목                 | Standard RAG               | LLM-Wiki                  
────────────────────────────────────────────────────────────────────────────
지식 표현              | 원본 청크 (raw text)           | LLM 정리 위키 (.md)           
지식 컴파일             | 없음 (매 쿼리 재발견)              | 1회 후 축적·재사용               
크로스 레퍼런스           | 없음                         | [[wiki links]] 자동 연결      
구축 비용              | 낮음 (임베딩만)                  | 높음 (논문당 LLM 호출)           
쿼리 속도              | 빠름 (벡터 검색)                 | 중간 (index 탐색)             
확장성                | 매우 높음 (FAISS)              | 중소 규모 적합                  
투명성                | 낮음 (청크 블랙박스)               | 높음 (.md 직접 열람)            
교차 질문 품질           | 낮음 (단편 청크)                 | 높음 (위키 종합)                

[Standard RAG 검색 통계]
  문서 3편 | 청크 178개 | FAISS 178벡터

  질문     대상                        Rank1 Score Rank1 출처                                
  -------------------------------------------

---

## 결론 및 고찰

### 재현 범위

| 모듈 | 활용 여부 | 비고 |
|------|----------|------|
| `embeddings.py` | ✅ | `EmbeddingEngine` — all-MiniLM-L6-v2 + FAISS 인덱스 |
| `standard_rag.py` | ✅ | `retrieve_only` + `query()` 4가지 질문으로 실행 |
| `karpathy_wiki.py` | ✅ | `KarpathyWiki` — 위키 컴파일 + 4가지 질문으로 쿼리 |
| `llm.py` | ✅ | OpenAI gpt-4o-mini 래퍼 (캐시 포함) |
| `evaluate.py` | ✅ | LLM-as-Judge 평가기 (코드 재현) |
| `utils.py` | ✅ | `chunk_text` 내부 사용 |
| `config.py` | ✅ | 설정값 관리 |

### 핵심 발견

1. **질문 언어 효과**: 한국어 질문(score 0.10~0.16) → 영어 질문으로 변경 시 유사도 상승.  
   `all-MiniLM-L6-v2`는 영어 최적화 모델이므로 영어 질문이 검색 품질에 직접 영향.

2. **질문 특이성 효과**: 논문 특정 질문(Q1~Q3)은 유사도 높음.  
   교차 질문(Q4)은 Standard RAG로 답하기 어려움 → LLM-Wiki가 유리한 케이스.

3. **두 방식의 금융 연구 적합성**:
   - 특정 논문 사실 확인, 대규모 문서 검색 → **Standard RAG**
   - 논문 간 개념 종합, 인용 추적, 연구 질문 탐색 → **Karpathy LLM-Wiki**